model/task_lifecycle.go

```diff
@@ -1570,24 +1570,36 @@ func resetStrandedTask(t *task.Task) error {
 		// If the task has already exceeded the unschedulable threshold, we
 		// don't want to restart it, so just mark it as finished.
 		if t.DisplayOnly {
-			for _, etID := range t.ExecutionTasks {
-				var execTask *task.Task
-				execTask, err := task.FindOneId(etID)
-				if err != nil {
+			var execTasks []task.Task
+			var err error
+
+			if t.ResetFailedWhenFinished && !t.ResetWhenFinished {
+				execTasks, err = task.Find(task.FailedTasksByIds(t.ExecutionTasks))
+			} else {
+				execTasks, err = task.FindAll(db.Query(task.ByIds(t.ExecutionTasks)))
+			}
+
+			if err != nil {
+				return errors.Wrap(err, "finding execution tasks")
+			}
+
+			if len(execTasks) == 0 {
+				return errors.Wrap(err, "no execution tasks found")
+			}
+
+			for _, execTask := range execTasks {
+				if err != nil { // Keeping existing error check logic?
 					return errors.Wrap(err, "finding execution task")
 				}
-				if execTask == nil {
-					return errors.New("execution task not found")
-				}
-				if err := MarkEnd(execTask, evergreen.MonitorPackage, time.Now(), &t.Details, false); err != nil {
+				if err := MarkEnd(&execTask, evergreen.MonitorPackage, time.Now(), &t.Details, false); err != nil {
 					return errors.Wrap(err, "marking execution task as ended")
 				}
 			}
 		}
 		return errors.WithStack(MarkEnd(t, evergreen.MonitorPackage, time.Now(), &t.Details, false))
 	}
 
-	return errors.Wrap(ResetTaskOrDisplayTask(t, evergreen.User, evergreen.MonitorPackage, false, &t.Details), "resetting task")
+	return errors.Wrap(ResetTaskOrDisplayTask(t, evergreen.User, evergreen.MonitorPackage, t.ResetFailedWhenFinished && !t.ResetWhenFinished, &t.Details), "resetting task")
```

Comment:
If you look at what ResetTaskOrDisplayTask does, it uses this boolean to determine whether to _set_ ResetFailedWhenFinished, or whether to set ResetWhenFinished. So we should just be setting this to true so that we toggle ResetFailedWhenFinished to true. If a user has already requested to restart the whole task, then ResetWhenFinished will also already be true and that'll be handled later.

Reply:
I forced it to true instead of this logic! I also tested for completed status in the loop above